File: web/src/modules/container/api/dashboard-summary.test.ts
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { request } from '@/utils/request';

import { CONTAINER_API_PATH } from '../contract/paths';
import { getContainerDashboardSummary, mapContainerDashboardSummary } from './dashboard-summary';

vi.mock('@/utils/request', () => ({
  request: {
    get: vi.fn(),
  },
}));

describe('container dashboard summary api', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('reads the canonical dashboard summary path', async () => {
    const requestGet = vi.mocked(request.get);
    requestGet.mockResolvedValueOnce({
      overview: {
        running_containers: 10,
        abnormal_containers: 2,
        cpu_total_percent: 18.5,
      },
      hotspots: {
        cpu_top: [],
        memory_top: [],
      },
      anomalies: [],
    } as never);

    await getContainerDashboardSummary();

    expect(requestGet).toHaveBeenCalledWith({
      url: CONTAINER_API_PATH.DASHBOARD_SUMMARY,
    });
  });

  it('maps the canonical dashboard summary response into the dashboard view model', () => {
    expect(
      mapContainerDashboardSummary({
        collected_at: '2026-06-24T12:03:00Z',
        overview: {
          running_containers: 8,
          abnormal_containers: 3,
          cpu_total_percent: 52.5,
          memory_total_usage_bytes: 268435456,
          memory_total_limit_bytes: 536870912,
          memory_total_percent: 61.2,
        },
        hotspots: {
          cpu_top: [
            {
              id: 'cpu-1',
              name: 'api',
              short_id: 'cpu-1',
              image: 'graft/api:latest',
              state: 'running',
              health: 'healthy',
              restart_count: 1,
              resource: {
                available: true,
                stats_available: true,
                cpu_percent: 48.1,
                memory_percent: 18.2,
                memory_usage_bytes: 1207959552,
                memory_limit_bytes: 2147483648,
                collected_at: '2026-06-24T12:00:00Z',
              },
            },
          ],
          memory_top: [
            {
              id: 'mem-1',
              name: 'worker',
              short_id: 'mem-1',
              image: 'graft/worker:latest',
              state: 'running',
              resource: {
                available: true,
                stats_available: true,
                memory_percent: 72.4,
                memory_usage_bytes: 268435456,
                memory_limit_bytes: 536870912,
                collected_at: '2026-06-24T12:01:00Z',
              },
            },
          ],
        },
        anomalies: [
          {
            id: 'bad-1',
            name: 'scheduler',
            short_id: 'bad-1',
            image: 'graft/scheduler:latest',
            state: 'restarting',
            status: 'Restarting',
            reason_code: 'state.restarting',
            reason_label: 'Restarting',
            resource: {
              available: true,
              stats_available: true,
              cpu_percent: 2.4,
              memory_percent: 12.3,
              collected_at: '2026-06-24T12:02:00Z',
            },
          },
        ],
      } as never),
    ).toEqual({
      overview: {
        abnormalContainers: 3,
        collectedAt: '2026-06-24T12:03:00Z',
        cpuTotalPercent: 52.5,
        memoryTotalLimitBytes: 536870912,
        memoryTotalPercent: 61.2,
        memoryTotalUsageBytes: 268435456,
        runningContainers: 8,
      },
      hotspots: {
        cpu: [
          {
            cpuPercent: 48.1,
            health: 'healthy',
            id: 'cpu-1',
            image: 'graft/api:latest',
            memoryLimitBytes: 2147483648,
            memoryPercent: 18.2,
            memoryUsageBytes: 1207959552,
            name: 'api',
            restartCount: 1,
            shortId: 'cpu-1',
            state: 'running',
            collectedAt: '2026-06-24T12:00:00Z',
          },
        ],
        memory: [
          {
            collectedAt: '2026-06-24T12:01:00Z',
            cpuPercent: null,
            health: null,
            id: 'mem-1',
            image: 'graft/worker:latest',
            memoryLimitBytes: 536870912,
            memoryPercent: 72.4,
            memoryUsageBytes: 268435456,
            name: 'worker',
            restartCount: null,
            shortId: 'mem-1',
            state: 'running',
          },
        ],
      },
      anomalies: [
        {
          collectedAt: '2026-06-24T12:02:00Z',
          cpuPercent: 2.4,
          health: null,
          id: 'bad-1',
          image: 'graft/scheduler:latest',
          memoryPercent: 12.3,
          memoryUsageBytes: null,
          memoryLimitBytes: null,
          name: 'scheduler',
          reasonCode: 'state.restarting',
          reasonLabel: 'Restarting',
          restartCount: null,
          shortId: 'bad-1',
          state: 'restarting',
          status: 'Restarting',
        },
      ],
    });
  });
});

File: server/modules/container/dashboard_summary_test.go
package container

import (
	"context"
	"testing"

	"graft/server/modules/container/terminal"
)

func TestBuildContainerDashboardSummary(t *testing.T) {
	t.Parallel()

	cpuHeavy := fakeSummary()
	cpuHeavy.ID = "c-1"
	cpuHeavy.Name = "cpu-heavy"
	cpuHeavy.ShortID = "c1"
	cpuHeavy.Resource = resourceWithCPUAndMemory(82.5, 512, 1024)

	memoryHeavy := fakeSummary()
	memoryHeavy.ID = "c-2"
	memoryHeavy.Name = "memory-heavy"
	memoryHeavy.ShortID = "c2"
	memoryHeavy.Resource = resourceWithCPUAndMemory(10, 2048, 4096)

	unhealthy := fakeSummary()
	unhealthy.ID = "c-3"
	unhealthy.Name = "unhealthy"
	unhealthy.ShortID = "c3"
	unhealthy.Health = containerHealthUnhealthy
	unhealthy.Resource = resourceWithCPUAndMemory(1, 128, 256)

	exited := fakeSummary()
	exited.ID = "c-4"
	exited.Name = "exited"
	exited.ShortID = "c4"
	exited.State = "exited"
	exited.Status = "Exited"
	exited.Resource = resourceWithCPUAndMemory(0, 64, 256)

	result := buildContainerDashboardSummary([]Summary{cpuHeavy, memoryHeavy, unhealthy, exited})

	if result.Overview.RunningContainers != 3 {
		t.Fatalf("expected 3 running containers, got %d", result.Overview.RunningContainers)
	}
	if result.Overview.AbnormalContainers != 2 {
		t.Fatalf("expected 2 abnormal containers, got %d", result.Overview.AbnormalContainers)
	}
	if len(result.Hotspots.CPUTop) != 3 || result.Hotspots.CPUTop[0].ID != "c-1" {
		t.Fatalf("expected cpu top ordered by cpu usage, got %#v", result.Hotspots.CPUTop)
	}
	if len(result.Hotspots.MemoryTop) != 3 || result.Hotspots.MemoryTop[0].ID != "c-2" {
		t.Fatalf("expected memory top ordered by memory usage, got %#v", result.Hotspots.MemoryTop)
	}
	if len(result.Anomalies) != 2 {
		t.Fatalf("expected only abnormal containers included, got %#v", result.Anomalies)
	}
	if result.Anomalies[0].ID != "c-3" {
		t.Fatalf("expected unhealthy anomaly first, got %#v", result.Anomalies)
	}
	if result.Anomalies[1].ID != "c-4" {
		t.Fatalf("expected exited anomaly before hotspot-only rows, got %#v", result.Anomalies)
	}
	if result.Overview.MemoryTotalPercent == nil {
		t.Fatalf("expected aggregated memory total percent")
	}
}

func TestBuildContainerDashboardSummaryExcludesNormalLoadFromAnomalies(t *testing.T) {
	t.Parallel()

	cpuHeavy := fakeSummary()
	cpuHeavy.ID = "cpu"
	cpuHeavy.Name = "cpu"
	cpuHeavy.Resource = resourceWithCPUAndMemory(90, 128, 256)

	memoryHeavy := fakeSummary()
	memoryHeavy.ID = "mem"
	memoryHeavy.Name = "mem"
	memoryHeavy.Resource = resourceWithCPUAndMemory(8, 4096, 8192)

	result := buildContainerDashboardSummary([]Summary{cpuHeavy, memoryHeavy})
	if len(result.Anomalies) != 0 {
		t.Fatalf("expected hotspot-only rows to stay out of anomalies, got %#v", result.Anomalies)
	}
}

func TestServiceDashboardSummaryUsesRuntimeList(t *testing.T) {
	t.Parallel()

	summaryA := fakeSummary()
	summaryA.ID = "dashboard-a"
	summaryA.Name = "dashboard-a"
	summaryA.ShortID = "da"
	summaryA.Resource = resourceWithCPUAndMemory(20, 100, 200)

	summaryB := fakeSummary()
	summaryB.ID = "dashboard-b"
	summaryB.Name = "dashboard-b"
	summaryB.ShortID = "db"
	summaryB.State = "restarting"
	summaryB.Resource = resourceWithCPUAndMemory(5, 50, 200)

	service, err := newTestService(containerServiceOptions{
		runtime:                 listOnlyRuntime{items: []Summary{summaryA, summaryB}},
		enabled:                 true,
		dangerousActionsEnabled: true,
		shellEnabled:            true,
		defaultTail:             defaultContainerLogsDefaultTail,
		maxTail:                 defaultContainerLogsMaxTail,
		authorizer:              fakeAuthorizer{},
	})
	if err != nil {
		t.Fatalf("new service: %v", err)
	}

	result, err := service.DashboardSummary(context.Background(), dashboardSummaryQuery{})
	if err != nil {
		t.Fatalf("dashboard summary: %v", err)
	}
	if len(result.Hotspots.CPUTop) != 2 {
		t.Fatalf("expected 2 cpu hotspots, got %d", len(result.Hotspots.CPUTop))
	}
	if len(result.Anomalies) != 1 {
		t.Fatalf("expected only abnormal containers in anomalies, got %d", len(result.Anomalies))
	}
	if result.Anomalies[0].ID != "dashboard-b" {
		t.Fatalf("expected restarting container to remain the only anomaly, got %#v", result.Anomalies)
	}
}

func resourceWithCPUAndMemory(cpu float64, usage int64, limit int64) ResourceSummary {
	memoryPercent := 0.0
	if limit > 0 {
		memoryPercent = (float64(usage) / float64(limit)) * 100
	}
	return ResourceSummary{
		Available:        true,
		StatsAvailable:   true,
		CPUPercent:       &cpu,
		MemoryUsageBytes: &usage,
		MemoryLimitBytes: &limit,
		MemoryPercent:    &memoryPercent,
	}
}

type listOnlyRuntime struct {
	items []Summary
}

func (r listOnlyRuntime) Info(context.Context) (RuntimeInfo, error) {
	return fakeRuntime{}.Info(context.Background())
}
func (r listOnlyRuntime) List(context.Context, ListQuery) ([]Summary, error) {
	return append([]Summary(nil), r.items...), nil
}
func (r listOnlyRuntime) Detail(context.Context, Ref) (Detail, error)  { return Detail{}, nil }
func (r listOnlyRuntime) Mounts(context.Context, Ref) ([]Mount, error) { return nil, nil }
func (r listOnlyRuntime) MountUsage(context.Context, Ref, string) (MountUsage, error) {
	return MountUsage{}, nil
}
func (r listOnlyRuntime) Logs(context.Context, Ref, LogQuery) (Logs, error) { return Logs{}, nil }
func (r listOnlyRuntime) Shell(context.Context, Ref, string) (terminal.Session, error) {
	return newStubTerminalSession(), nil
}
func (r listOnlyRuntime) Start(context.Context, Ref) (ActionResult, error) {
	return ActionResult{}, nil
}
func (r listOnlyRuntime) Stop(context.Context, Ref) (ActionResult, error) { return ActionResult{}, nil }
func (r listOnlyRuntime) Restart(context.Context, Ref) (ActionResult, error) {
	return ActionResult{}, nil
}
func (r listOnlyRuntime) Remove(context.Context, Ref, RemoveOptions) (ActionResult, error) {
	return ActionResult{}, nil
}
func (r listOnlyRuntime) Close() error { return nil }
